frame/vamm/src/tests/swap_asset.rs

```diff
@@ -0,0 +1,477 @@
+
+proptest! {
+	#![proptest_config(ProptestConfig::with_cases(RUN_CASES))]
+	#[test]
+	fn swap_invalid_vamm_error(
+		vamm_state in get_vamm_state(Default::default()),
+		swap_config in get_swap_config(Default::default()),
+	) {
+		prop_assume!(swap_config.vamm_id != 0);
+
+		ExtBuilder {
+			vamm_count: 1,
+			vamms: vec![(0, vamm_state)]
+		}.build().execute_with(|| {
+			assert_noop!(
+				TestPallet::swap(&swap_config),
+				Error::<MockRuntime>::VammDoesNotExist
+			);
+		})
+	}
+}
+
+proptest! {
+	#[test]
+	fn fails_to_swap_assets_if_vamm_is_closed(
+		mut vamm_state in get_vamm_state(Default::default()),
+		mut swap_config in get_swap_config(Default::default()),
+		(close, now) in then_and_now()
+	) {
+		// Make the current time be greater than the time when the vamm is
+		// set to close, doing this we ensure we can't make swaps due to the
+		// vamm be closed.
+		vamm_state.closed = Some(close);
+		swap_config.vamm_id = 0;
+
+		ExtBuilder {
+			vamm_count: 1,
+			vamms: vec![(0, vamm_state)]
+		}.build().execute_with(|| {
+			run_to_block(now);
+
+			assert_noop!(
+				TestPallet::swap(&swap_config),
+				Error::<MockRuntime>::VammIsClosed
+			);
+		})
+	}
+}
+
+proptest! {
+	#![proptest_config(ProptestConfig::with_cases(RUN_CASES))]
+	#[test]
+	fn fails_to_swap_assets_if_output_is_less_than_minimum_limit(
+		mut vamm_state in any_vamm_state(),
+		mut swap_config in get_swap_config(Default::default()),
+		limit in balance_range_upper_half(),
+	) {
+		// Ensure vamm is open before start operation to swap assets.
+		vamm_state.closed = None;
+
+		// Ensure input amount will not cause `InsufficientFundsForTrade`,
+		// `Overflow`, `Underflow`, etc.
+		swap_config.input_amount = 0;
+
+		swap_config.output_amount_limit = limit;
+		swap_config.vamm_id = 0;
+
+		ExtBuilder {
+			vamm_count: 1,
+			vamms: vec![(0, vamm_state)]
+		}.build().execute_with(|| {
+			assert_noop!(
+				TestPallet::swap(&swap_config),
+				Error::<MockRuntime>::SwappedAmountLessThanMinimumLimit
+			);
+		})
+	}
+}
+
+proptest! {
+	#![proptest_config(ProptestConfig::with_cases(RUN_CASES))]
+	#[test]
+	fn swap_asset_suceeds_emitting_event(
+		mut vamm_state in get_vamm_state(Default::default()),
+		mut swap_config in get_swap_config(Default::default()),
+		base in balance_range_low(),
+		quote in balance_range_low(),
+		peg in balance_range_low(),
+	) {
+		// Ensure vamm is open before start operation to swap assets.
+		vamm_state.closed = None;
+
+		// Ensure values won't overflow when computing swap.
+		vamm_state.base_asset_reserves = base;
+		vamm_state.quote_asset_reserves = quote;
+		vamm_state.peg_multiplier = peg;
+
+		// Disable output limit check
+		swap_config.output_amount_limit = Balance::zero();
+
+		swap_config.vamm_id = 0;
+
+		ExtBuilder {
+			vamm_count: 1,
+			vamms: vec![(0, vamm_state)]
+		}.build().execute_with(|| {
+			// For event emission
+			run_to_block(1);
+
+			let swap = TestPallet::swap(&swap_config);
+
+			if let Ok(swap) = swap {
```

Comment:
shouldn't you call `ensure_ok!(swap)` first and then call `system::assert_last_event(...)`

Reply:
You're correct, but then we stumble up in the error we're facing with some specific swaps (the error I'm yet to tackle next week), so this was the way I found to test that events were being emitted correct despite some specific swaps failing. 

I should've added a TODO comment about changing it later tho, sorry about it.